core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java

```diff
@@ -0,0 +1,322 @@
+            double skewness = ((double) (maxWriterWrittenBytes - minWriterWrittenBytes)) / maxWriterWrittenBytes;
+            if (skewness <= SKEWNESS_THRESHOLD || isNaN(skewness)) {
+                break;
+            }
+
+            minSkewedWriters.add(new EntryWithPriority<>(minWriter.value(), minWriterWrittenBytes));
+        }
+        return minSkewedWriters.build();
+    }
+
+    private Optional<IdWithWrittenBytes> rebalancePartition(
+            int minWriterId,
+            long maxWriterWrittenBytesSinceLastRebalance,
+            long maxWriterRowCountSinceLastRebalance,
+            IndexedPriorityQueue<Integer> maxWriterPartitions,
+            Set<Integer> rebalancedPartitionIds)
+    {
+        while (true) {
+            // Find the partition with maximum row count
+            EntryWithPriority<Integer> partitionToScale = maxWriterPartitions.pollEntryWithPriority();
+
+            if (partitionToScale == null) {
+                return Optional.empty();
+            }
+
+            int partitionId = partitionToScale.value();
+            long rowCount = partitionToScale.priority();
+
+            PartitionWriterInfo partitionWriterInfo = partitionWriterInfos[partitionId];
+
+            // If partition is already balanced or min skewed writer is already writing to that partition, then skip
+            // this partition and move on to the other partition inside max writer.
+            if (rebalancedPartitionIds.contains(partitionId) || partitionWriterInfo.containsWriter(minWriterId)) {
+                continue;
+            }
+
+            long estimatedPartitionWrittenBytes = (maxWriterWrittenBytesSinceLastRebalance * rowCount) / maxWriterRowCountSinceLastRebalance;
+
+            // Scale the partition when estimated physicalWrittenBytes is greater than writerMinSize.
+            if (partitionWriterInfo.getWriterCount() <= numberOfWriters && estimatedPartitionWrittenBytes >= writerMinSize) {
+                partitionWriterInfo.addWriter(minWriterId);
+                log.warn(format("Scaled partition (%s) to writer %s with writer count %s", partitionId, minWriterId, partitionWriterInfo.getWriterCount()));
+                return Optional.of(new IdWithWrittenBytes(partitionId, estimatedPartitionWrittenBytes));
+            }
+
+            return Optional.empty();
+        }
+    }
+
+    private void resetStateForNextRebalance(long[] writerPhysicalWrittenBytes)
+    {
+        for (int i = 0; i < numberOfWriters; i++) {
+            writerPartitionIdWithRowCount.set(i, new IndexedPriorityQueue<>());
+            lastRefreshWriterPhysicalWrittenBytes[i] = writerPhysicalWrittenBytes[i];
+        }
+    }
+
+    public synchronized void addPartitionRowCount(int writerId, int partitionId, long rowCount)
+    {
+        IndexedPriorityQueue<Integer> partitions = writerPartitionIdWithRowCount.get(writerId);
+        Long currentRowCount = partitions.getPriority(partitionId);
```

Comment:
Could we keep partition row count in `ParitionStats` of some kind instead of exposing additional (internal) capabilities of `IndexedPriorityQueue`?

We can even keep a field like `long[] partitionRowCounts` in `UnifornPartitionRebalancer`.